cpp/include/cudf/join.hpp

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace cudf {
+// joins
+
+/** 
+ * @brief  Performs an inner join on the specified columns of two
+ * dataframes (left, right)
+ *
+ * If join_context->flag_method is set to GDF_SORT then the null_count of the
+ * columns must be set to 0 otherwise @throws GDF Validity is unsupported by sort_join.
+ * 
+ * @example TableA a:{0,1,2}
+ *          TableB b:{1,2,3}, c:{4,5,6} And TableB joins TableA 
+ *          on column 'a' with column 'b'.
+ * The result would be Table-res a:{1, 2}, b:{1, 2}, c:{4, 5}
+ * Result is intersection of column 'a' and column 'b'
+ *
+ * @param[in] left_table The left dataframe
+ * @param[in] right_table The right dataframe
+ * @param[in] join_ind_table The table containing two columns,
+ * first column - indices of left table provided for join
+ * second column - indices of right table provided for join
+ * @param[in] merging_join_ind_table Similar to join_ind_table,
+ * contains two columns, but these column have the indices
+ * which have same name in dataframes and will eventually merge
+ * into a single column.
+ * @param[out] out_index_table The table containing 
+ * joined indices of left and right table 
+ * @param[in] join_context The context to use to control how 
+ * the join is performed,e.g., sort vs hash based implementation
+ * 
+ * @returns If Success, pair of left and right table which are merged 
+ */
+std::pair<cudf::table, cudf::table> inner_join(
+                         cudf::table const& left_table,
+                         cudf::table const& right_table,
+                         cudf::table const& join_ind_table,
+                         cudf::table const& merging_join_ind_table,
+                         cudf::table const& out_index_table,
+                         gdf_context *join_context);
+/** 
+ * @brief  Performs a left join (also known as left outer join) on the
+ * specified columns of two dataframes (left, right)
+ *
+ * If join_context->flag_method is set to GDF_SORT then the null_count of the
+ * columns must be set to 0 otherwise @throws GDF Validity is unsupported by sort_join.
+ * 
+ * @example TableA a:{0,1,2}
+ *          TableB b:{1,2,3}, c:{4,5,6} And TableB joins TableA 
+ *          on column 'a' with column 'b'.
+ * The result would be Table-res a:{0, 1, 2}, b:{nan, 1, 2}, c:{nan, 4, 5}
```

Comment:
NaN and NULL are not the same thing. 
```suggestion
 * The result would be Table-res a:{0, 1, 2}, b:{NULL, 1, 2}, c:{NULL, 4, 5}
```